hoomd/md/PotentialPairAlchemical.h

```diff
@@ -0,0 +1,719 @@
+// Copyright (c) 2009-2022 The Regents of the University of Michigan.
+// Part of HOOMD-blue, released under the BSD 3-Clause License.
+
+#pragma once
+
+#include <bitset>
+#include <cstddef>
+#include <memory>
+#include <numeric>
+
+#include "hoomd/md/PotentialPair.h"
+
+/*! \file PotentialPairAlchemical.h
+    \brief Defines the template class for alchemical pair potentials
+*/
+
+#ifdef __HIPCC__
+#error This header cannot be compiled by nvcc
+#endif
+
+namespace hoomd
+    {
+
+namespace md
+    {
+
+template<class evaluator> struct AlchemyPackage
+    {
+    bool calculate_derivatives = false;
+    std::vector<std::array<Scalar, evaluator::num_alchemical_parameters>> alphas = {};
+    std::vector<ArrayHandle<Scalar>> force_handles = {};
+    std::vector<std::bitset<evaluator::num_alchemical_parameters>> compute_mask = {};
+
+    AlchemyPackage(std::nullptr_t) {};
+    AlchemyPackage() {};
+    };
+
+//! Template class for computing alchemical pair potentials
+/*! <b>Overview:</b>
+
+    <b>Implementation details</b>
+
+
+
+    \sa export_PotentialPair()
+*/
+template<class evaluator,
+         typename extra_pkg = AlchemyPackage<evaluator>,
+         typename alpha_particle_type = AlchemicalPairParticle>
+class PotentialPairAlchemical : public PotentialPair<evaluator>
+    {
+    public:
+    //! Construct the pair potential
+    PotentialPairAlchemical(std::shared_ptr<SystemDefinition> sysdef,
+                            std::shared_ptr<NeighborList> nlist);
+    //! Destructor
+    virtual ~PotentialPairAlchemical();
+
+    std::shared_ptr<alpha_particle_type> getAlchemicalPairParticle(int i, int j, int k)
+        {
+        std::shared_ptr<alpha_particle_type>& alpha_p
+            = m_alchemical_particles[k * m_alchemy_index.getNumElements() + m_alchemy_index(i, j)];
+        if (alpha_p == nullptr)
+            {
+            alpha_p = std::make_shared<alpha_particle_type>(m_exec_conf, make_int3(i, j, k));
+            }
+        return alpha_p;
+        }
+
+    void enableAlchemicalPairParticle(std::shared_ptr<alpha_particle_type> alpha_p)
+        {
+        m_alchemy_mask[m_alchemy_index(alpha_p->m_type_pair_param.x, alpha_p->m_type_pair_param.y)]
+                      [alpha_p->m_type_pair_param.z]
+            = true;
+        alpha_p->resizeForces(m_pdata->getN());
```

Comment:
I am not sure why this line is here.

Reply:
It appears that `AlchemicalMDParticle` has 0-length arrays on construction. This call is needed to initialize the arrays.